stdlib/public/core/Zip.swift

```diff
@@ -136,13 +147,140 @@ extension Zip2Sequence.Iterator: IteratorProtocol {
+///
+///     let numbers = 0...2
+///     let emptyString = ""
+///     let zipped = zipLongest(numbers, emptyString).lazy
+///
+///     print(Array(zipped.map({($0.0 ?? 0, $0.1 ?? "a")})))
+///     // Prints "[(0, "a"), (1, "a"), (2, "a")]"
+///
+/// The second example compares two ranges for equality:
+///
+///     let shorter = 0...1
+///     let longer = 0...2
+///
+///     print(zipLongest(shorter, longer).lazy.allSatisfy(==))
+///     // Prints "false"
+///
+/// `zip(_:_:)` would have given us a false positive:
+///
+///     print(zip(shorter, longer).lazy.allSatisfy(==))
+///     // Prints "true"
+///
+/// The third example merges two ranges containing integers into an array by
+/// adding their respective (default) elements:
+///
+///     print(zipLongest(shorter, longer).lazy.map({($0.0 ?? 0) + ($0.1 ?? 0)}))
+///     // Prints "[0, 2, 2]"
+///
+/// Composing `map(_:)` with a zip-like operation is commonly known as `zipWith`
+/// in other languages.
+///
+/// - Note: Use `zip(_:_:)` if you don't require the latter elements of the
+/// longest sequence or if the length of both sequences are equal.
+///
+/// - Parameters:
+///   - sequence1: The first `Sequence`.
+///   - sequence2: The second `Sequence`.
+///
+/// - Returns: An instance of `ZipLongest2Sequence`.
+@inlinable // FIXME(sil-serialize-all)
+public func zipLongest<Sequence1, Sequence2>(
+  _ sequence1: Sequence1, _ sequence2: Sequence2
+) -> ZipLongest2Sequence<Sequence1, Sequence2> {
+  return ZipLongest2Sequence(_sequence1: sequence1, _sequence2: sequence2)
+}
+
+/// A sequence of pairs created from a pair of sequences by zipping their
+/// elements until the *longest* sequence is exhausted, padding the latter
+/// elements of the shortest sequence with `nil`.
+///
+/// - Note: Instances of `ZipLongest2Sequence` can be created with the global
+/// `zipLongest(_:_:)` function.
+@_fixed_layout // FIXME(sil-serialize-all)
+public struct ZipLongest2Sequence<Sequence1 : Sequence, Sequence2 : Sequence> {
+  @usableFromInline // FIXME(sil-serialize-all)
+  internal let _sequence1: Sequence1
+  @usableFromInline // FIXME(sil-serialize-all)
+  internal let _sequence2: Sequence2
+
+  /// Constructs an instance of `ZipLongest2Sequence`.
+  @inlinable // FIXME(sil-serialize-all)
+  public // @testable
+  init(_sequence1 sequence1: Sequence1, _sequence2 sequence2: Sequence2) {
+    (_sequence1, _sequence2) = (sequence1, sequence2)
+  }
+}
+
+extension ZipLongest2Sequence {
+  /// The iterator for `ZipLongest2Sequence`.
+  ///
+  /// - Note: To create an instance, use `makeIterator()`.
+  @_fixed_layout // FIXME(sil-serialize-all)
+  public struct Iterator {
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _baseStream1: Sequence1.Iterator
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _baseStream2: Sequence2.Iterator
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _reachedEnd: Bool = false
```

Comment:
Unlike with regular `zip`, I don't think it's necessary to track this. You can just rely on the two iterators.

Reply:
👍